kv/kvnemesis: TestKVNemesisSingleNode failed [locktable nil txn/guard]

kv/kvnemesis.TestKVNemesisSingleNode [failed](https://tanzanite.cluster.engflow.com/invocation/17e905b9-37a7-43a3-82fa-f3ebd7bd394f) on master @ [414cb46513079bb8869bd776f045ed6c96ad6d26](https://github.com/cockroachdb/cockroach/commits/414cb46513079bb8869bd776f045ed6c96ad6d26):


```
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*keyLocks).enqueueLockingRequest
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table.go:2733
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*keyLocks).scanAndMaybeEnqueue
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table.go:2386
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockTableGuardImpl).resumeScan
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table.go:931
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*lockTableImpl).ScanAndEnqueue
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock_table.go:3855
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).sequenceReqWithGuard
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:335
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency.(*managerImpl).SequenceReq
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/concurrency_manager.go:239
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:468
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:185
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:193
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:202
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1333
        	            	  | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
        	            	  | 	github.com/cockroachdb/cockroach/pkg/server/node.go:1464
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.func1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:704
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:815
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:815
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:105
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.makeInternalClientAdapter.chainUnaryServerInterceptors.bindUnaryServerInterceptorToHandler.func4
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:815
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136
        	            	Wraps: (2) secondary error attachment
        	            	  | runtime error: invalid memory address or nil pointer dereference
        	            	  | (1) runtime error: invalid memory address or nil pointer dereference
        	            	  | Error types: (1) runtime.errorString
        	            	Wraps: (3) panic applying step 
        	            	  | db0.ScanForShare(ctx, tk(1854443122230575953), tk(17482680577027038983), 0): runtime error: invalid memory address or nil pointer dereference
        	            	Error types: (1) *withstack.withStack (2) *secondary.withSecondaryError (3) *errutil.leafError
    panic.go:523: -- test log scope end --
test logs left over in: outputs.zip/logTestKVNemesisSingleNode4096416147
--- FAIL: TestKVNemesisSingleNode (4.13s)
```
<p>Parameters: <code>attempt=1</code>
, <code>run=27</code>
, <code>shard=1</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33259

--- a/pkg/kv/kvserver/concurrency/lock_table.go
+++ b/pkg/kv/kvserver/concurrency/lock_table.go
@@ -3386,29 +3386,28 @@ func (kl *keyLocks) recomputeWaitQueues(st *cluster.Settings) {
 			break
 		}
 		removed := false
-		if qlr.active {
-			// A queued locking request, that's actively waiting, no longer conflicts
-			// with locks on this key -- it can be allowed to proceed. There's two
-			// cases:
-			// 1. If it's a transactional request, it needs to acquire a claim by
-			// holding its place in the lock wait queue while marking itself as
-			// inactive.
-			// 2. Non-transactional requests do not acquire claims, so they can be
-			// removed from the wait queue.
-			if qlr.guard.txn == nil {
-				kl.removeLockingRequest(curr)
-				removed = true
-			} else {
-				qlr.active = false // mark as inactive
-				if qlr.guard == kl.distinguishedWaiter {
-					// A new distinguished waiter will be selected by informActiveWaiters.
-					kl.distinguishedWaiter = nil
-				}
-				qlr.guard.mu.Lock()
-				qlr.guard.doneActivelyWaitingAtLock()
-				qlr.guard.mu.Unlock()
+		if qlr.guard.txn == nil {
+			// We're dealing with a queued non-transactional locking request that no
+			// longer conflicts with locks/claims on this key. As non-transactional
+			// requests do not acquire claims, the request can be removed from the
+			// wait queue.
+			kl.removeLockingRequest(curr)
+			removed = true
+		} else if qlr.active {
+			// A transactional locking request that no longer conflicts with
+			// locks/claims on this key but is actively waiting. It no longer needs to
+			// do so, but before it can proceed, it must acquire a (possibly joint)
+			// claim. It does so by marking itself as inactive.
+			qlr.active = false // mark as inactive
+			if qlr.guard == kl.distinguishedWaiter {
+				// A new distinguished waiter will be selected by informActiveWaiters.
+				kl.distinguishedWaiter = nil
 			}
+			qlr.guard.mu.Lock()
+			qlr.guard.doneActivelyWaitingAtLock()
+			qlr.guard.mu.Unlock()
 		}
+
 		// Locking requests conflict with both the lock holder(s) and other lower
 		// sequence numbered locking requests in the lock's wait queue, so we may
 		// need to update strongestMode before moving on with our iteration.

--- a/pkg/kv/kvserver/concurrency/testdata/lock_table/shared_locks
+++ b/pkg/kv/kvserver/concurrency/testdata/lock_table/shared_locks
@@ -1320,6 +1320,95 @@ num=1
     active: false req: 68, strength: Shared, txn: 00000000-0000-0000-0000-000000000003
     active: false req: 69, strength: Shared, txn: 00000000-0000-0000-0000-000000000004
 
+# ------------------------------------------------------------------------------
+# Regression test for
+# https://github.com/cockroachdb/cockroach/issues/113924#issuecomment-1799057694.
+# ------------------------------------------------------------------------------
+
+clear
+----
+num=0
+
+new-request r=req70 txn=txn1 ts=10 spans=intent@a
+----
+
+scan r=req70
+----
+start-waiting: false
+
+new-request r=req71 txn=none ts=10 spans=intent@a
+----
+
+scan r=req71
+----
+start-waiting: false
+
+add-discovered r=req71 txn=txn2 k=a durability=u strength=shared
+----
+num=1
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 0, iso: Serializable, info: repl [Shared]
+   queued locking requests:
+    active: false req: 71, strength: Intent, txn: none
+
+scan r=req70
+----
+start-waiting: true
+
+new-request r=req72 txn=txn3 ts=10 spans=shared@a
+----
+
+scan r=req72
+----
+start-waiting: true
+
+new-request r=req73 txn=none ts=10 spans=shared@a
+----
+
+scan r=req73
+----
+start-waiting: true
+
+new-request r=req74 txn=none ts=10 spans=intent@a
+----
+
+scan r=req74
+----
+start-waiting: true
+
+print
+----
+num=1
+ lock: "a"
+  holder: txn: 00000000-0000-0000-0000-000000000002 epoch: 0, iso: Serializable, info: repl [Shared]
+   queued locking requests:
+    active: true req: 70, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 71, strength: Intent, txn: none
+    active: true req: 72, strength: Shared, txn: 00000000-0000-0000-0000-000000000003
+    active: true req: 73, strength: Shared, txn: none
+    active: true req: 74, strength: Intent, txn: none
+   distinguished req: 70
+
+release txn=txn2 span=a
+----
+num=1
+ lock: "a"
+   queued locking requests:
+    active: false req: 70, strength: Intent, txn: 00000000-0000-0000-0000-000000000001
+    active: false req: 71, strength: Intent, txn: none
+    active: true req: 72, strength: Shared, txn: 00000000-0000-0000-0000-000000000003
+    active: true req: 73, strength: Shared, txn: none
+    active: true req: 74, strength: Intent, txn: none
+   distinguished req: 72
+
+dequeue r=req70
+----
+num=1
+ lock: "a"
+   queued locking requests:
+    active: false req: 72, strength: Shared, txn: 00000000-0000-0000-0000-000000000003
+    active: true req: 74, strength: Intent, txn: none
+   distinguished req: 74
 
 # TODO(arul): (non-exhaustive list) of shared lock state transitions that aren't
 # currently supported (and we need to add support for):

